src/io/convert/gdf-to-csr.cu

```diff
@@ -0,0 +1,322 @@
+ *
+ */
+
+#include <gdf/gdf.h>
+#include <gdf/errorutils.h>
+#include "gdf/gdf_io.h"
+
+#include <thrust/scan.h>
+#include <thrust/execution_policy.h>
+
+using namespace std;
+
+//--- all the private functions
+template<typename T>
+gdf_error runConverter(gdf_column **gdfData, csr_gdf *csrReturn, int64_t * offsets);
+
+
+//--- private CUDA functions / kernels
+template<typename T>
+__global__ void cudaCreateCSR(void *data, gdf_valid_type *valid, gdf_dtype dtype, int colID, T *A, int64_t *JA, int64_t *offsets, gdf_size_type numRows);
+
+__global__ void determineValidRecCount(gdf_valid_type *validArray, gdf_size_type numRows, gdf_size_type numCol, int64_t * offset);
+
+template<typename T>
+__device__ T convertDataElement(gdf_column *gdf, int idx, gdf_dtype dtype);
+
+__device__ int whichBitmapCSR(int record) { return (record/8);  }
+__device__ int whichBitCSR(int bit) { return (bit % 8);  }
+__device__ int checkBitCSR(gdf_valid_type data, int bit) {
+
+	gdf_valid_type bitMask[8] 		= {1, 2, 4, 8, 16, 32, 64, 128};
+	return (data & bitMask[bit]);
+}
+
+
+//
+//------------------------------------------------------------
+//
+
+/*
+ * Convert a Dense GDF into a CSR GDF
+ *
+ * Restrictions:  All columns need to be of the same length
+ */
+/**
+ * @brief convert a GDF into a CSR
+ *
+ * Take a matrix in GDF format and convert it into a CSR.  The column major matrix needs to have every column defined.
+ * Passing in a COO datset will be treated as a two column matrix
+ *
+ * @param gdfData the ordered list of columns
+ * @param numCol the number of columns in the gdfData array
+ * @param csrReturn a pointer to the returned data structure
+ *
+ */
+gdf_error gdf_to_csr(gdf_column **gdfData, int numCol, csr_gdf *csrReturn) {
+
+	int64_t			numNull = 	0;
+	int64_t			nnz		= 	0;
+	gdf_size_type	numRows =	gdfData[0]->size;
+	gdf_dtype		dType	=	gdf_dtype::GDF_invalid;		// the data type to make the CSR element array (A)
+
+	/**
+	 * Currently the gdf_dtype enum is arranged based on data size, as long as it stays that way the enum values can be
+	 * exploited by just picking the largest enum value
+	 *
+	 * While looping, also get the number of null values (this will work one day)
+	 */
+	for ( int x =0; x < numCol; x++) {
+		if( gdfData[x]->dtype > dType)
+			dType = gdfData[x]->dtype;
+
+		numNull += gdfData[x]->null_count;
+	}
+
+	if (dType == gdf_dtype::GDF_invalid || dType == gdf_dtype::GDF_STRING )
+		return gdf_error::GDF_UNSUPPORTED_DTYPE;
+
+	// the number of valid elements is simple the max number of possible elements (rows * columns) minus the number of nulls
+	// the current problem is that algorithms are not setting null_count;
+	// gdf_size_type is 32bits (int) but the total size could be larger than an int, so use a long
+	nnz = (numRows * numCol) - numNull;
+
+	// Allocate space for the offset - this will eventually be IA - dtype is long since the sum of all column elements could be larger than int32
+    long * offsets;
```

Comment:
you mean "long long"? also, why not size_t because offsets will never be bigger than addressable memory.

Reply:
yes, it should be either int64_t or size_t